### PR TITLE
[Feature] Optimize indexer

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

This PR is used to improve https://github.com/ray-project/kuberay/pull/946. This PR description will discuss the reasons why I make the change and why I do not make the change. This may be more memory-efficient (I do not have a chance to benchmark it.)

### Indexer
This PR is to optimize indexer (cache) based on kubebuilder's tutorial ([tutorial](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html), [code](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/internal/controller/cronjob_controller.go#L556-L572)).  In #946, we create a local cache for all Pods, including Ray Pods and other Pods. Currently, we only index Ray Pods.

### resource event handler (not update)

* See [here](https://github.com/ray-project/kuberay/pull/946#discussion_r1150095430) for more context.

Initially, I thought that we should implement all filter logic in the resource event handler (i.e. `CreateFunc`) and return true only when an event is issued by a Ray Pod. With this approach, we can make the function `eventReconcile` only handle event processing and not worry about event filtering. Therefore, the label `common.RayNodeLabelKey` is not necessary.

However, I have since realized that the owners of the unhealthy events are Pods, which may include Ray Pods and other Pods. Therefore, we cannot use `GetControllerOf` to get the owner and determine whether the event belongs to a Ray Pod or not by checking whether the owner's kind is "RayCluster" or not.

Hence, the implementation will be more complex. If we want to distinguish whether an event belongs to a Ray Pod or not, we need to

* Get the UID of the event's InvolvedObject.
* Use the UID to look up Pods from the local cache.
* Use `GetControllerOf(pod)` to get the owner, and check whether the kind is RayCluster or not.

It will make the function eventReconcile easier to understand, but we need to have more I/O (read local cache) which may cause performance overhead. In summary, we should use `common.RayNodeLabelKey` to distinguish Ray Pods in eventReconcile until we understand the cost of the look-up operation.

### Fake client limitation

The fake client included in controller-runtime does not support the use of `client.MatchingFields` with ListOptions. For more details, see https://github.com/kubernetes-sigs/controller-runtime/issues/866. We need to figure out all limitations of the fake client in the future, and may be we should figure out a better solution.

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

Pass the test in #946.

### CODE REVIEW

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -46,15 +46,19 @@ var (
 	PrioritizeWorkersToDelete bool
 	ForcedClusterUpgrade      bool
 	EnableBatchScheduler      bool
-
-	// Definition of a index field for pod name
-	podUIDIndexField = "metadata.uid"
+	PodUIDIndexField          = "metadata.uid"
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:49]: Oops, I forget that. Thank you for pointing this out! By the way, is this PR necessary to be included in the v0.5.0 or it is not that urgent?

[the remark this answers is not in the document]
Comment (kevin85421) [reply ray-operator/controllers/ray/raycluster_controller.go:49]: I prefer not to include this in v0.5.0 except this optimization can prevent OOM.

[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:49]: > I prefer not to include this in v0.5.0 except this optimization can prevent OOM.

I am ok with both. This is not a big change.

[the remark this answers is not in the document]
Comment (wilsonwang371) [reply ray-operator/controllers/ray/raycluster_controller.go:49]: You can go ahead and make the change and i will give a pass.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 2]
```diff
@@ -46,15 +46,19 @@ var (
 	PrioritizeWorkersToDelete bool
 	ForcedClusterUpgrade      bool
 	EnableBatchScheduler      bool
-
```
Comment (wilsonwang371) [inline ray-operator/controllers/ray/raycluster_controller.go:49]: change PodUIDIndexField to podUIDIndexField

**Review by wilsonwang371 — CHANGES_REQUESTED** [review]
please update the variable and we can merge them after this.

**Comment by kevin85421** [comment]
Thank you for the follow-up. I will update this PR after I finish the v0.5.0-rc.0 release process. In addition, this PR will not be included in v0.5.0 because I did not have a chance to benchmark the memory usage.

**Pull request closed without merging**